### PR TITLE
TFP: haku near-winner replication seed=7

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Replicate the haku near-winner TFP config (best_val=0.002466, best_test=0.002393) with seed=7 to sample the result distribution across seeds. Part of a 4-seed sweep (seeds 0, 7, 42, 123) to determine whether the config reliably lands below the historical target of field_mse < 0.002383.

## Dataset

`tandemfoil_paper` — paper-faithful Experiment 4 tasks, normalized full-field MSE.

## Training Command

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1e-4 \
  --cosine-t-max 20 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --model-slices 64 \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-fourier \
  --ema-decay 0.999 \
  --epochs 999 \
  --seed 7 \
  --wandb_group tfp-haku-replication
```

## Baseline

- **best_val field_mse:** 0.002466 (haku W&B y790y0dv)
- **best_test field_mse:** 0.002393
- **Historical target:** 0.002383
- **Kill gates:** field_mse > 0.006 at ep220, > 0.004 at ep350, no improvement for 160 epochs

## Expected Outcome

Confirm stability of the haku config at seed=7. A result at or below 0.002383 would beat the historical target. Report best_val and best_test (from best-val checkpoint) field_mse alongside the W&B run ID.

## Instructions

1. Run the training command above exactly as written.
2. Monitor: kill if field_mse > 0.006 at ep220, > 0.004 at ep350, or no improvement for 160 consecutive epochs.
3. Report best_val/field_mse, best_test/field_mse (evaluated from best-val checkpoint), and the W&B run ID in a PR comment.
4. Include the full W&B run URL so the advisor can verify metrics independently.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT wolfwood:

## Results

**Run killed at epoch 220 — kill gate triggered (best field_mse 0.014247 > 0.006 threshold)**

### Metrics (best checkpoint)

| Metric | Value | Baseline (haku near-winner) | Delta |
|--------|-------|----------------------------|-------|
| **val_primary/field_mse** | **0.014247** | 0.002466 | +478% (5.8x worse) |
| val/surface_mse | 0.009060 | — | — |
| val/surface_mse_p | 0.017809 | — | — |
| val/volume_mse | 0.014329 | — | — |
| best epoch | 128 | — | — |

No test metrics — process killed before final evaluation.

### Training Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 1e-4 \
  --cosine-t-max 20 --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 \
  --model-heads 3 --enable-fourier --ema-decay 0.999 \
  --epochs 999 --seed 7 --agent wolfwood \
  --wandb-name "wolfwood/tfp-haku-seed7" \
  --wandb-group tfp-haku-replication
```

### W&B

- **Run ID:** jgwtzegj
- **URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/jgwtzegj

### What happened

The run diverged catastrophically around epoch 130. After a healthy descent phase (field_mse dropping from ~0.36 at ep1 to 0.014247 at ep128), the model entered sustained NaN gradient territory. From ep130 onward, current field_mse oscillated wildly (0.3–1.1) and eventually settled into a flat diverged regime (~0.117) from ep200+. The EMA checkpoint preserved the ep128 best but the model never recovered.

The best result of 0.014247 is 5.8x worse than the haku near-winner target of 0.002466. At seed=7, this config does **not** replicate the near-winner performance — it hits the ep220 kill gate decisively.

**Key observation:** This config lacks the physics features (TE coord frame, Cp panel, asinh pressure, residual prediction) that the haku champion (#3025) used. The W&B config shows `enable_te_coord_frame=False`, `enable_cp_panel=False`, `asinh_pressure=False`, `residual_prediction=False`. The only feature augmentation is Fourier encoding. This may explain the performance gap — the physics features may be essential for reaching the 0.002–0.003 field_mse regime on TFP.

### Suggested follow-ups

1. **Verify the original haku near-winner config:** The PR command may be missing physics flags that the actual near-winner used. Worth cross-referencing with haku's W&B run y790y0dv to confirm the exact config.
2. **Re-run seed=7 with physics features:** If the near-winner config should have included physics features, re-run with `--enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition`.
3. **T_max=20 stability:** The divergence at ep130 with T_max=20 may be a stability issue specific to this config. The champion used T_max=10 with similar divergence at ep462 — T_max=20 may accelerate the instability.

**Comment by tcapelle** [comment]
ADVISOR: Config error identified — the original assignment was missing ALL physics features (TE coord frame, Cp panel, asinh pressure, residual prediction, pressure prior). This explains the 5.8x worse result. Good catch noting this in your results!

**Please re-run with the CORRECT full TFP champion config (T_max=10, all physics features, seed=7):**

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion \
  --lr 1e-4 \
  --cosine-t-max 10 \
  --grad-clip 0.5 \
  --weight-decay 1e-2 \
  --model-layers 3 \
  --model-hidden-dim 192 \
  --model-heads 3 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --ema-decay 0.999 \
  --epochs 999 \
  --seed 7 \
  --wandb-group tfp-champion-replication \
  --kill-if-best-val-above 220:0.006 \
  --kill-if-best-val-above 350:0.004 \
  --kill-if-no-improvement 160:0.001
```

**Baseline to beat:** val_primary/field_mse = 0.002344 (PR #3056, haku). Target: lower is better. Report test_primary/field_mse from best checkpoint.

**Comment by tcapelle** [comment]
STUDENT wolfwood:

## Results (Re-run: Full Champion Config with Physics Features, seed=7)

**Run killed at epoch 171 — no-improvement kill gate triggered (best field_mse 19,618 at ep11, no improvement for 160 epochs)**

### Metrics (best checkpoint, ep11)

| Metric | Value | Baseline (haku champion #3056) | Delta |
|--------|-------|-------------------------------|-------|
| **val_primary/field_mse** | **19,618** | 0.002344 | 8.4M x worse |
| **test_primary/field_mse** | **360,817** | — | — |
| test/surface_mse | 31,426 | — | — |
| test/surface_mse_Ux | 0.00621 | — | — |
| test/surface_mse_Uy | 0.01951 | — | — |
| test/surface_mse_p | 94,279 | — | — |
| test/volume_mse | 366,056 | — | — |
| best epoch | 11 | — | — |

### Training Command

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset tandemfoil_paper --optimizer lion --lr 1e-4 \
  --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only --asinh-pressure \
  --residual-prediction --enable-pressure-prior-addition \
  --ema-decay 0.999 --epochs 999 --seed 7 --agent wolfwood \
  --wandb-name "wolfwood/tfp-champion-seed7" \
  --wandb-group tfp-champion-replication \
  --kill-if-best-val-above 220:0.006 \
  --kill-if-best-val-above 350:0.004 \
  --kill-if-no-improvement 160:0.001
```

### W&B

- **Run ID:** e59108sw
- **URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/e59108sw

### What happened

The model **catastrophically diverged** from the very first epoch and never recovered. Key observations:

1. **Epoch 1-3:** Rapid early descent from 6.3e+28 → 19,618 (typical for residual prediction)
2. **Epoch 4+:** Violent oscillations began — field_mse swung between millions and billions, with spikes as high as 129 trillion (ep77)
3. **Brief recovery attempts:** Around ep54-57 (down to 192K) and ep114-126 (down to 156K), the model briefly descended before spiking again
4. **Final regime (ep130-171):** Settled into persistent 100K-60M oscillations with no convergence

**This is dramatically worse than Run 1 (no physics features), which achieved val field_mse = 0.014247.** The full champion config with physics features somehow performs 1.4 million times worse on this seed.

### Analysis: Why did physics features make it worse?

Key differences from Run 1:
- **model_slices:** 96 (default) vs 64 in Run 1 — the original near-winner PR specified 64
- **cosine T_max:** 10 vs 20 in Run 1
- **Physics features:** all enabled (TE coord frame, Cp panel, asinh pressure, residual prediction, pressure prior)

The most likely explanation is that **the combination of physics features + residual prediction creates a much harder optimization landscape** at seed=7. The residual prediction pre-subtraction with asinh-pressure transformed targets may amplify gradient instability early on, and the Lion optimizer at lr=1e-4 with T_max=10 doesn't recover from the initial divergence.

It's also possible the **model_slices=96 (vs 64) matters** — the original near-winner config used 64, and this wasn't included in the corrected command.

### Suggested follow-ups

1. **Re-run with model_slices=64:** The original near-winner used `--model-slices 64`. This was in the first PR command but dropped in the corrected one. It may be load-bearing.
2. **Cross-reference haku champion W&B config (PR #3056):** Verify the exact model_slices, and whether any other flags differ from what was specified.
3. **Try seed=7 with lower lr (e.g. 5e-5):** The instability pattern suggests the learning rate may be too aggressive for this seed with all features enabled.
4. **Consider that seed=7 is a bad seed for this config:** The original (non-physics) config also failed at seed=7 (best 0.014247 vs target 0.002383). This seed may be inherently unstable for the TFP task.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3291 because ADVISOR: Closing — seed=7 is catastrophically unstable for TFP with physics features (field_mse=19,618, 8.4 million times worse than baseline). The first run (no physics, seed=7) also failed (0.014247). Seed=7 is a bad seed for TFP regardless of config.

Note: the model_slices question is a red herring — the haku champion (#3056) also used the default (no --model-slices flag). The TFP champion command in BASELINE.md doesn't specify model_slices.

Three other TFP seed runs are in flight with corrected configs (#3287 seed=0, #3288 seed=42, #3292 seed=123). Wolfwood reassigned to DM..
